nomad/structs/structs.go

```diff
@@ -5000,6 +5006,50 @@ func (tg *TaskGroup) validateNetworks() error {
 	return mErr.ErrorOrNil()
 }
 
+// validateServices runs Service.Validate() on group-level services,
+// checks that group services do not conflict with task services and that
+// group service checks that refer to tasks only refer to tasks that exist.
+func (tg *TaskGroup) validateServices() error {
+	var mErr multierror.Error
+	knownTasks := make(map[string]struct{})
+	knownServices := make(map[string]struct{})
+
+	// Create a map of known tasks and their services so we can compare
+	// vs the group-level services and checks
+	for _, task := range tg.Tasks {
+		knownTasks[task.Name] = struct{}{}
+		if task.Services == nil {
+			continue
+		}
+		for _, service := range task.Services {
+			if _, ok := knownServices[service.Name+service.PortLabel]; ok {
+				mErr.Errors = append(mErr.Errors, fmt.Errorf("service %q is duplicate", service.Name))
+			}
+			knownServices[service.Name+service.PortLabel] = struct{}{}
+		}
+	}
+	for i, service := range tg.Services {
+		if err := service.Validate(); err != nil {
+			outer := fmt.Errorf("service[%d] %+q validation failed: %s", i, service.Name, err)
+			mErr.Errors = append(mErr.Errors, outer)
+		}
+		if _, ok := knownServices[service.Name+service.PortLabel]; ok {
+			mErr.Errors = append(mErr.Errors, fmt.Errorf("service %q is duplicate", service.Name))
+		}
+		knownServices[service.Name+service.PortLabel] = struct{}{}
+
+		for _, check := range service.Checks {
+			if check.Task != nil {
```

Comment:
Do only certain kinds of `ServiceCheck` need/support having an associated `Task`? i.e. should we be restricting this to script checks only?

Reply:
Yeah I think its a validation error if you set `check.Task` when it is a non-script check. :+1:

Reply:
Nope! It will be needed for HTTP/gRPC checks once Consul 1.6.1 ships. Check Restart stanzas need to know what task to restart.

Reply:
Fixed.